Allow access to restricted system indices for reserved system roles

williamrandolph opened the pull request:
After we merged https://github.com/elastic/elasticsearch/pull/74212 , the Kibana team saw permissions errors at startup.

The cause of this is that all system indices are now restricted indices from the perspective of security, when previously the only restricted indices were security and async search indices. In that PR, we didn't update the reserved system roles so that the ones that access system indices have `allowRestrictedIndices = true`.

We can find the roles that need this change by adding system index patterns to the `TestRestrictedIndices` class and looking for failures in `ReservedRolesStoreTests`.

The fixes for the particular kibana issue are the updates to the `kibana_system` role.

williamrandolph committed:
Broadly allow access to restricted indices from the Kibana role

elasticmachine commented:
Pinging @elastic/es-security (Team:Security)

elasticmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

williamrandolph committed:
Merge branch 'master' into kibana-role-allow-restricted-indices

tylersmalley commented:
I verified this resolved the issue preventing the promotion of the nightly snapshot for Kibana.

ywangd commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java:
```diff
@@ -383,18 +383,20 @@ public static RoleDescriptor kibanaSystemRoleDescriptor(String name) {
                 // Data telemetry reads mappings, metadata and stats of indices
                 RoleDescriptor.IndicesPrivileges.builder()
                     .indices("*")
-                    .privileges("view_index_metadata", "monitor").build(),
+                    .privileges("view_index_metadata", "monitor")
+                    .allowRestrictedIndices(true).build(),
```
This change allows the `kibana_system` user to have some access to the `.security` index. I don't think this is what we want.

williamrandolph replied:
I went through the list of system indices and pared down where I was allowing restricted indices to just the places where we are dealing with system indices, so I hope I've addressed this concern.

ywangd replied:
Yes the change addressed my previous comment, thanks. But we still have another semantic change: before the #74212, this index privilege allow kibana to `monitor` and `view_index_metadata` for indices like `.watches`, `.triggered_watches`. These indices are now no longer accessible for `kibana_system`. Maybe this is the right thing to do. But I'd like to double check whether this is intentional.

tvernum commented:
It looks to me like this issue slipped through because `TestRestrictedIndices` only contains security and async search indices.

Can we update it to include the complete set of out-of-the-box system indices?

williamrandolph committed:
Merge branch 'master' into kibana-role-allow-restricted-indices

williamrandolph committed:
Roll back broad permissions

williamrandolph committed:
Add system index patterns to TestRestrictedIndices

A missing piece in #74212 was system index patterns in the tests for the
ReservedRolesStore. Without these patterns, the tests did not accurately
check whether a role was incorrectly accessing a system index that was
not previously a restricted index.

This commit adds all of the current system index patterns to the test
class and adds restricted index access to the system roles that need it
for tests to pass.

williamrandolph commented:
@tvernum 

Thanks for the pointer. I think that is exactly the issue. I've manually added all the system index patterns to the `TestRestrictedIndices` class. I would really like to refactor this work so that we are referencing the actual string constants that the system index plugins use, but I think there's a good argument for pushing that to a follow-up and merging a fix for the Kibana team as soon as possible so that they can update their Elasticsearch snapshots.

ywangd reviewed (commented):
The way how privileges are granted to restricted indices feels brittle to me. It requires checking and matching names from different places carefully (as well as adding them to tests). But maybe it is because this PR is a one-time mass update. It may feel alright for future smaller changes. Also I don't have a good suggestion for alternatives. 

I think we need a conclusion on `kibana_system`'s `monitor` and `view_index_metadata` access for `*`. Should we strictly maintain the previous semantic or are we ok to change? I am ok either way as long as we are conscious on the decision. Thanks!

williamrandolph committed:
Perserve existing Kibana data telemetry privileges

williamrandolph committed:
Merge branch 'master' into kibana-role-allow-restricted-indices

ywangd reviewed (approved):
LGTM

> I think it could be argued that we should exclude other system indices from data telemetry

I tend to agree especially this change is 8.0 only. But I am not familiar with all system indices and how they should be surfaced by Kibana. I think it needs input from people who are more familiar with its usage (maybe a round of emails?). For this PR, I think maintaining the exact same semantic is the safest. 

Can I also suggest we add a few more assertions about `kibana_system` cannot view metadata or monitor `.security*` and `.async-search*`? The [current assertions](https://github.com/elastic/elasticsearch/blob/d36f24fbc367c3c51be81ba1c8481d0bd7b5a95e/x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStoreTests.java#L599-L600) only ensure no read or write access to the indices, but not metadata level access. I think this is also why the test didn't tripped on the first version of this PR.

williamrandolph committed:
Test that data telemetry can't access security and async indices
